pennylane/devices/qubit/preprocess.py

```diff
@@ -359,20 +394,32 @@ def preprocess(
             options for the execution.
 
     Returns:
-        Tuple[QuantumTape], Callable, ExecutionConfig: QuantumTapes that the device can natively execute,
-        a postprocessing function to be called after execution, and a configuration with originally unset specifications filled in.
+        TransformProgram, ExecutionConfig: A transform program and a configuration with originally unset specifications filled in.
     """
-    for c in circuits:
-        validate_measurements(c, execution_config)
+    transform_program = TransformProgram()
 
-    circuits = tuple(expand_fn(c) for c in circuits)
+    # Validate measurement
+    valid_meas_transform = TransformContainer(
+        transform=validate_measurements.transform, args=[execution_config]
+    )
+    transform_program.push_back(valid_meas_transform)
+
+    # Circuit expand
+    expand_transform = TransformContainer(transform=expand_fn.transform)
+    transform_program.push_back(expand_transform)
+
+    # Adjoint expand
     if execution_config.gradient_method == "adjoint":
-        circuits = tuple(validate_and_expand_adjoint(c) for c in circuits)
-        for circuit_or_error in circuits:
-            if isinstance(circuit_or_error, DeviceError):
-                raise circuit_or_error  # it's an error
+        expand_adjoint_transform = TransformContainer(
+            transform=validate_and_expand_adjoint.transform
+        )
+        transform_program.push_back(expand_adjoint_transform)
 
-    transform = partial(batch_transform, execution_config=execution_config)
-    circuits, batch_fn = qml.transforms.map_batch_transform(transform, circuits)
+    ### Broadcast expand
+    if circuits[0].batch_size is not None and execution_config.gradient_method == "adjoint":
+        broadcast_expand_transform = TransformContainer(
+            transform=qml.transforms.broadcast_expand.transform
+        )
+        transform_program.push_back(broadcast_expand_transform)
 
-    return circuits, batch_fn, _update_config(execution_config)
+    return transform_program, _update_config(execution_config)
```

Comment:
I kinda liked how `dev.preprocess(tape)[0][0]` was a tape, made it very easy to quickly see what a device does most of the time while still allowing for more advanced stuff. Now, I'd have to do `dev.preprocess(tape)[0](tape)[0][0]`, right? Not a dealbreaker for me, but I do wanna make sure.

While I wrote that out, it made me wonder if we should be able to keep the tapes somewhere on the transform program (ie. the tapes used to define the program) so we don't need to pass them around multiple times. again, just curious about your thoughts